Refresh line

garethspor opened the pull request:
Hey guys!

Commit:
LinenoiseRemotePrint: remotely pass text to print

When other processes called printf and linenoise had a prompt displayed,
the text got printed after the prompt and looked horrible. This function
restarts the line before printing with a \r, prints, then pushes a
ctrl-r to a special pipe that gets waited on alongside stdin. That
ctrl-r now triggers the refresh line function. The result is your text
is printed, then the prompt + incomplete input is redrawn below.

Commit:
linenoiseRemoteRefreshLine

Tidied some code and switched the function to a simpler verison that
simply refreshes the line.

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +822,44 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+            }
+            else
+            {
+                continue;
```
When does this happen? Since you only pass readfds wouldn't retval being non-zero indicate one of them is set? Should this be a FATAL_ERROR? At least a comment?

garethspor replied:
adding a comment...

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +822,44 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+        retval = select(nfds, &rfds, NULL, NULL, NULL);
+        if (retval == -1)
+        {
+            fprintf(stderr, "Problem with select in linenoise\n");
```
add strerror(errno) 
also... maybe this should be fatal?

yromanenko commented on linenoise.c:
```diff
@@ -757,6 +761,18 @@ void linenoiseEditDeletePrevWord(struct linenoiseState *l) {
+        return;
+    }
+    int ctrl_pipe_write_fd = ctrl_pipe_fd[1];
+    char buf[1] = {18}; //ctrl-r;
```
That's my bad but you have CTRL_R right there, so just use that

groman2 commented on linenoise.c:
```diff
@@ -794,10 +828,49 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+        int nread = 0;
         char seq[3];
 
-        nread = read(l.ifd,&c,1);
```
Now that's a read!

groman replied:
not awesome at all...

groman2 reviewed (commented):
Awesome!

de45 commented:
test comment

Commit:
RemoteRefreshLine: several small changes

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +822,46 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+        else
+        {
+            // timeout, try again :)
+            continue;
```
So yeah next step for this commit is to make this delayed
I recommend adding a flag called "delayedRefresh"
and right here:
```c
if(delayedRefresh)
{
   refreshLine(&l);
}
```

yromanenko commented on linenoise.c:
```diff
@@ -944,6 +1008,9 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
             linenoiseEditDeletePrevWord(&l);
             break;
+        case CTRL_R: /* ctrl-r, refresh current line */
+            refreshLine(&l);
```
Here, instead of calling refreshLine() directly set `delayedRefresh = 1`

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +822,46 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+        FD_SET(ctrl_pipe_read_fd, &rfds);
+        int nfds = (l.ifd > ctrl_pipe_read_fd) ? l.ifd+1 : ctrl_pipe_read_fd+1; //compute 1+(the greater fd) for select
+
+        retval = select(nfds, &rfds, NULL, NULL, NULL);
```
```c
retval = select(nfds, &rfds, NULL, NULL, delayedRefresh ? &timeout : NULL)
```
or something like that

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +822,46 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+            // timeout, try again :)
+            continue;
+        }
+
```
```c
delayedRefresh = 0;
```

yromanenko reviewed (commented):
Added suggestions of how to do delayed refresh

Commit:
Delayed refresh

When a refresh line request comes in, register it but don't refresh the
line until there's been a short timeout. This should prevent multiple
requests from piling up.

Commit:
removed errant continue

special commented on linenoise.c:
```diff
@@ -132,6 +132,9 @@ static int history_max_len = LINENOISE_DEFAULT_HISTORY_MAX_LEN;
 static char **history = NULL;
 
+static int ctrl_pipe_created = 0;
+static int ctrl_pipe_fd[2] = {0, 0};
```
nitpick: `{-1,-1}` might be better default values, because 0 is a valid fd.

yromanenko replied:
Agree

special commented on linenoise.c:
```diff
@@ -757,6 +761,18 @@ void linenoiseEditDeletePrevWord(struct linenoiseState *l) {
+void linenoiseRemoteRefreshLine()
+{
+    if (!ctrl_pipe_created)
+    {
```
nitpick: this file puts braces on the same line, and omits braces for single line statements

special reviewed (approved):
Seems generally good

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +824,54 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+
+        struct timeval timeout;
+        timeout.tv_sec = 0;
+        timeout.tv_usec = 1e5; // 0.1 seconds
```
`1e5` is a floating point literal (a double) which I felt like commenting on but will be converted implicitly by the compiler and I'm really far into this sentence and still can't think of any reason this could be problematic.

yromanenko commented on linenoise.c:
```diff
@@ -794,10 +824,54 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
+            }
+            continue;
+        }
+
```
Should you have a `delayedRefresh = 0` here as well, otherwise typing at the prompt will not prevent it from happening just put it off until you're done typing really fast...

Commit:
Code cleanup
